handlers_test.go

```diff
@@ -91,3 +131,89 @@ func Test_GivenCorrectId_HandlerReturnsDeviceObject(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, expected, result)
 }
+
+func Test_PageAndLimitWrapper_GivenWrongInput_HandlerReturns400(t *testing.T) {
+	r := newRouter(NewService(&mockDao{}))
+	mockServer := httptest.NewServer(r)
+
+	tests := map[string]string{
+		"limit is not int":               "/devices?limit=a",
+		"page is not int":                "/devices?page=a",
+		"page is not int, correct limit": "/devices?limit=1&page=a",
+		"limit is below zero":            "/devices?limit=-1",
+		"page is below zero":             "/devices?page=-1",
+	}
+	for name, tc := range tests {
+		resp, err := http.Get(mockServer.URL + tc)
+
+		assert.NoError(t, err, name)
+		assert.Equal(t, http.StatusBadRequest, resp.StatusCode, name)
+	}
+}
+
+func Test_PageAndLimitWrapper_NoParams_HandlerDefaultsLimitAndPage(t *testing.T) {
+	r := newRouter(NewService(&mockDao{}))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices")
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+}
+
+func Test_GetManyDevicesHandler_GivenDaoError_HandlerReturns500(t *testing.T) {
+	r := newRouter(NewService(&mockDao{returnErr: ErrDao("")}))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices")
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+}
+
+func Test_GetManyDevicesHandler_GivenLimitAndPage_HandlerReturnsCorrectNumberOfDevicesPerPage(t *testing.T) {
+	t.Skip()
```

Comment:
Another skipped test - either fix the implementation or the test.

Additionally, you test here the behaviour of the handler and the service.
Tests on handler level should test only handler responsibilities. In this case handler is responsible for:
- reading page and limit
- calling right service method with read page and limit
- returning results or error